### PR TITLE
WFLY-9648 Add ability to configure non-blocking state transfer for replicated/distributed caches

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.jboss.org/browse/EAP7-888
https://issues.jboss.org/browse/WFLY-9648

### CODE REVIEW

File: clustering/infinispan/WFLY-9648_Initial_State_Transfer.adoc [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+=== Nice-to-Have Requirements
+
+=== Non-Requirements
+
+== Test Plan
+
+Run standard set of failover smartfrog tests with state-transfer configured with a 0 timeout.
+Validate that async, non-blocking state transfer results does not introduce sampling errors.
+Generate visual feedback of performance impact of the additional remote gets following failover.
```
Comment (mvinkler) [inline clustering/infinispan/WFLY-9648_Initial_State_Transfer.adoc]: So you would like to have a performance test with a failover?

Comment (rhusar) [reply clustering/infinispan/WFLY-9648_Initial_State_Transfer.adoc]: Surely having a performance test for the events where failover happens is needed. Ideally these should be the same as the original failover tests we have but obtain better representation of performance data.

Comment (mvinkler) [reply clustering/infinispan/WFLY-9648_Initial_State_Transfer.adoc]: I guess we can run a standard failover test and collect statistics as we do in performance tests, output is a set of graphs (throughput, response time, etc.) you are familiar with.

Comment (pferraro) [reply clustering/infinispan/WFLY-9648_Initial_State_Transfer.adoc]: @mvinkler Correct.

**Comment by mvinkler** [comment]
QE feedback: looks OK.

File: clustering/infinispan/WFLY-9648_Initial_State_Transfer.adoc (hunk 1 above)
Comment (tommaso-borgato) [reply clustering/infinispan/WFLY-9648_Initial_State_Transfer.adoc]: @pferraro 
`/subsystem=infinispan/cache-container=ejb/distributed-cache=dist/component=state-transfer:add(enabled=true,timeout=0)`
works correctly but outputs the following warning:
`12:23:05,030 INFO  [org.jboss.as.controller.management-deprecated] (pool-3-thread-1) WFLYCTL0028: Attribute 'enabled' in the resource at address '/subsystem=infinispan/cache-container=ejb/distributed-cache=dist/component=state-transfer' is deprecated, and may be removed in a future version. See the attribute description in the output of the read-resource-description operation to learn more about the deprecation.
{"outcome" => "success"}`
the same for:
`/subsystem=infinispan/cache-container=server/replicated-cache=default/component=state-transfer:add(enabled=true,timeout=0)
12:25:41,056 INFO  [org.jboss.as.controller.management-deprecated] (pool-3-thread-2) WFLYCTL0028: Attribute 'enabled' in the resource at address '/subsystem=infinispan/cache-container=server/replicated-cache=default/component=state-transfer' is deprecated, and may be removed in a future version. See the attribute description in the output of the read-resource-description operation to learn more about the deprecation.
{"outcome" => "success"}`

Comment (pferraro) [reply clustering/infinispan/WFLY-9648_Initial_State_Transfer.adoc]: @tommaso-borgato I'm no longer allowing state-transfer to be disabled - as this has unintended consequences.  This RFE has been limited to just enabling non-blocking initial state-transfer, via timeout=0.

Comment (tommaso-borgato) [reply clustering/infinispan/WFLY-9648_Initial_State_Transfer.adoc]: @pferraro: for the stress test with the state-transfer disabled, the configuration is `<state-transfer enabled="false"/>` correct?

**Comment by jamezp** [comment]
This was merged back in March and it looks like this got missed.

**Pull request merged**